Lib/test/test_frame.py

```diff
@@ -431,6 +432,116 @@ def test_is_mapping(self):
                 kind = "other"
         self.assertEqual(kind, "mapping")
 
+    def test_proxy_key_stringlikes(self):
+        class StringSubclass(str):
+            pass
+
+        class ImpostorX:
+            def __hash__(self):
+                return hash('x')
+
+            def __eq__(self, other):
+                return other == 'x'
+
+        def f(obj):
+            x = 1
+            proxy = sys._getframe().f_locals
+            proxy[obj] = 2
+            return (
+                proxy.keys() ==  ['obj', 'x', 'proxy'],
+                proxy == {'obj': 'x', 'x': 2, 'proxy': proxy},
+                proxy
+            )
+
+        for obj in StringSubclass('x'), ImpostorX():
+            with self.subTest(cls=type(obj).__name__):
+
+                assertion1, assertion2, proxy = f(obj)
+                self.assertEqual(proxy.keys(),  ['obj', 'x', 'proxy'])
+                self.assertEqual(proxy, {'obj': 'x', 'x': 2, 'proxy': proxy})
+                self.assertTrue(assertion1)
+                self.assertTrue(assertion2)
```

Comment:
I think this test case will be clearer if it takes value snapshots inside the function, and then leaves the comparisons outside where they can use the `unittest` rich assertion methods:

```suggestion
        def f(obj):
            x = 1
            proxy = sys._getframe().f_locals
            proxy[obj] = 2
            return (
                list(proxy.keys()),
                dict(proxy),
                proxy
            )

        for obj in StringSubclass('x'), ImpostorX():
            with self.subTest(cls=type(obj).__name__):

                keys_snapshot, proxy_snapshot, proxy = f(obj)
                expected_keys = ['obj', 'x', 'proxy']
                expected_dict = {'obj': 'x', 'x': 2, 'proxy': proxy}
                self.assertEqual(proxy.keys(),  expected_keys)
                self.assertEqual(proxy, expected_dict)
                self.assertEqual(keys_snapshot,  expected_keys)
                self.assertEqual(proxy_snapshot, expected_dict)
```